components/org.wso2.carbon.appmgt.mdm.osgiconnector/src/main/java/org/wso2/carbon/appmgt/mdm/osgiconnector/ApplicationOperationsImpl.java

```diff
@@ -0,0 +1,251 @@
+		mobileApp.setPlatform(app.getPlatform());
+		mobileApp.setVersion(app.getVersion());
+		Properties properties = new Properties();
+
+		if (MDMAppConstants.IOS.equals(app.getPlatform())) {
+			if (MDMAppConstants.ENTERPRISE.equals(app.getType())) {
+				properties.put(MDMAppConstants.IOSConstants.IS_REMOVE_APP, true);
+				properties.put(MDMAppConstants.IOSConstants.IS_PREVENT_BACKUP, true);
+			} else if (MDMAppConstants.IOSConstants.PUBLIC.equals(app.getType())) {
+				properties.put(MDMAppConstants.IOSConstants.I_TUNES_ID,
+				               Integer.parseInt(app.getIdentifier()));
+				properties.put(MDMAppConstants.IOSConstants.IS_REMOVE_APP, true);
+				properties.put(MDMAppConstants.IOSConstants.IS_PREVENT_BACKUP, true);
+			} else if (MDMAppConstants.WEBAPP.equals(app.getType())) {
+				properties.put(MDMAppConstants.IOSConstants.LABEL, app.getName());
+				properties.put(MDMAppConstants.IOSConstants.IS_REMOVE_APP, true);
+			}
+		} else if (MDMAppConstants.WEBAPP.equals(app.getPlatform())) {
+			properties.put(MDMAppConstants.IOSConstants.LABEL, app.getName());
+			properties.put(MDMAppConstants.IOSConstants.IS_REMOVE_APP, true);
+		}
+		mobileApp.setProperties(properties);
+		try {
+			for (DeviceIdentifier deviceIdentifier : deviceIdentifiers) {
+				if (deviceIdentifier.getType().equals(Platform.android.toString())) {
+					if (MDMAppConstants.INSTALL.equals(applicationOperationAction.getAction())) {
+						operation = AndroidApplicationOperationUtil
+								.createInstallAppOperation(mobileApp);
+					} else {
+						operation = AndroidApplicationOperationUtil
+								.createAppUninstallOperation(mobileApp);
+					}
+				} else if (deviceIdentifier.getType().equals(Platform.ios.toString())) {
+					if (MDMAppConstants.INSTALL.equals(applicationOperationAction.getAction())) {
+						operation =
+								IOSApplicationOperationUtil.createInstallAppOperation(mobileApp);
+					} else {
+						operation =
+								IOSApplicationOperationUtil.createAppUninstallOperation(mobileApp);
+					}
+				}
+				MDMServiceAPIUtils.getAppManagementService(applicationOperationAction.getTenantId())
+				                  .installApplicationForDevices(operation, deviceIdentifiers);
+			}
+		} catch (DeviceApplicationException mdmExce) {
+			log.error("Error in creating operation object using app.", mdmExce);
+			throw new MobileApplicationException(mdmExce);
+		} catch (ApplicationManagementException appMgtExce) {
+			log.error("Error in app installation.", appMgtExce);
+			throw new MobileApplicationException(appMgtExce);
+		}
+
+	}
+
+	/**
+	 * Create a new device identifier from Device object.
+	 * @param device device which is to be retrieved type and id
+	 * @return created device identifier
+	 */
+	private static DeviceIdentifier getDeviceIdentifierByDevice(
+			org.wso2.carbon.device.mgt.common.Device device) {
+		DeviceIdentifier deviceIdentifier = new DeviceIdentifier();
+		deviceIdentifier.setId(device.getDeviceIdentifier());
```

Comment:
Started a discussion in dev mailing list regarding this under the subject "Why we maintain different kind of device identifiers in device management core". Once the thread concludes, it should be properly fixed.